---
title: "Problem Set 3. Due Thurs March 2 5pm"
author: "Casey Martin"
date: "March 1, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment = "#>")
```

# Submission

1. Fork this repository to your own account
1. Make changes to the `README.Rmd` file (including the author field above).
1. Knit the file to HTML format and add it to your git repository (`git add`)
1. Submit a pull request with your Rmd and knitted HTML files.

# Overview 

You will examine a [time-course of mRNA expression][1] of barley powdery mildew fungus in immunocompromised plants (*Arabidopsis*). The fungus Blumeria graminis f. sp. hordei (called *Bgh* from now on)

Counts were generated using the [`maSigPro` software][3] and deposited in [NCBI GEO][2].

[1]: http://www.ncbi.nlm.nih.gov/pubmed/23175786
[2]: http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE43163  
[3]: http://bioinformatics.oxfordjournals.org/content/30/18/2598.full

Some important identifiers for the data set:

| Name | Description |
| ---- | ----------- |
| `A6` | *Bgh* isolate expressing other AVRA effectors |
| `K1` | *Bgh* isolate expressing the cognate AVRA1 effector for MLA1|
| `pps` | *Arabidopsis* plants |
| `B12` | *Arabidopsis* plants expressing MLA1-HA|

We will download tables from GEO containing counts of mRNA abundance from both the fungal sample.

# Raw data

First we need load the data.

```{r load_raw_data, message = FALSE}
library(tidyverse)
library(stringr)
library(ggplot2)

# bgh data
bgh_url <- 'http://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE43163&format=file&file=GSE43163_CompleteCountTable_Bgh.txt.gz'

# 2 info and blank lines at top, skip them
raw_data <- read_tsv(bgh_url, skip = 2)

# the header for the first column is "NA", set it manually
names(raw_data)[1] <- 'gene.name'

raw_data <- as_data_frame(raw_data)
```

# Problems


## Problem 1


The key steps are to `gather()` the data into key-value pairs and then `separate()` the information from the key column into new columns.

```{r problem_1}
raw_data %>% 
  gather(raw_data, value, -gene.name) %>%
  separate(raw_data, c('plant', 'fungus', 'time.point', 'rep'), sep = "_") -> tidy_data

myanswer1 = tidy_data

print(myanswer1)
```

## Problem 2

You need to translate what you learned on the command line into the R world.

Which plant has the highest expression of any gene in the 6hpi time point?

```{bash, eval=FALSE}
#raw_data | awk '$4 == "6hpi"' | sort -k6nr | head -n 1 | cut -f2
```

```{r}
tidy_data %>%
  filter(time.point == '6hpi') %>%
  arrange(desc(value)) -> tidy_sorted6h

myanswer2a = tidy_sorted6h[1,2]

print(myanswer2a)
```

Which plant / fungus pair has the highest expression in the 18hpi time point?

```{bash, eval=FALSE}
#raw_data | awk '$4 == "18hpi"' | sort -k6nr | head -n 1 | cut -f2,3
```

```{r}
tidy_data %>%
  filter(time.point == '18hpi') %>%
  arrange(desc(value)) -> tidy_sorted18h

myanswer2b= tidy_sorted6h[1, 2:3]

print(myanswer2b)
```

## Problem 3

Identify the top 3 most consistently differentially expressed genes between the earliest and latest time points for each combination of `plant` and `fungus` strains.

- "Differential expression"" is the difference between `value` (i.e., gene expression level) between time points.

- "Consistency" is the smallest variance in `value` between replicates.

### Strategy

1. Create a new table from the cleaned data by moving each `hpi` value to a new column name, with counts for each in the column (hint: use a `tidyr` verb). It is helpful to reformat the `hpi` values by converting from e.g. `6hpi` to `hpi.6`. You can use `mutate` to do this, i.e.:

```{r reformat_time, eval=TRUE}
tidy_data %>%
  mutate(time.value = str_replace(time.point, 'hpi', ''),
    time.hpi = str_c('hpi.', time.value)) -> tidy_data
```

2. Create a new column containing the expression difference between the relevant time points.

3. Calculate summary statistics (mean and variance) of the expression differences by grouping (hint) the `gene.name`, `plant`, and `virus` columns.

4. Sort by these statistics and use the `dplyr` verb `slice` to pull the ones you want (i.e., the top 3). Note you will have to remove `gene.name` from the grouping so that sorting works.

```{r problem_3}

### returns normalized foldchange for a given timepoint
foldchange_normal <- function(x, mytime.hpi)  {
  x %>%
    filter(time.hpi == "hpi.6" | time.hpi == mytime.hpi) %>%
    group_by(gene.name, plant, fungus, rep) %>%
    summarize(foldchange = log((value[time.hpi == mytime.hpi] + 0.01)/(value[time.hpi == "hpi.6"] + 0.01), 2)) %>%
    group_by(gene.name, plant, fungus) %>%
    summarize(foldchange_mean = mean(foldchange), foldchange_abs = abs(foldchange_mean),
              foldchange_var = var(abs(foldchange))) %>%
    mutate(time.hpi = mytime.hpi) -> mytidy_foldchange
  return(mytidy_foldchange)
}
mytimepoints = (unique(tidy_data$time.hpi)[-1])

### for loops are ugly, but I couldn't figure out why sapply returns a list of distorted tibbles
# sapply(mytimepoints, function(x) foldchange_normal(tidy_data, x)))
tidy_foldchange = data_frame()
for (i in 1:length(mytimepoints)){
  tidy_foldchange = bind_rows(tidy_foldchange, foldchange_normal(tidy_data, mytimepoints[i]))
}

# sets threshhold for subsetting top 1% mean foldchange_abs.
top_quantile = function(myarray, myquantile){
return(rev(quantile(myarray, probs = seq(0, 1, myquantile)))[2])
}

# subset the top 1% mean foldchange_abs and sort by lowest foldchange_var
tidy_foldchange %>%
  arrange(desc(foldchange_abs)) %>%
  group_by(gene.name, plant, fungus) %>%
  filter(foldchange_abs >= top_quantile(foldchange_abs, 0.01)) %>%
  arrange(foldchange_var) %>%
  group_by(plant, fungus) %>%
  slice(1:3) -> top_consistent

myanswer3 = top_consistent
  

print(myanswer3)
```

## Problem 4

The above table highlights the top 3 most consistently differentially expressed genes in pairwise infections of *Arabidopsis* (`r unique(myanswer3$plant)`) and *Bgh* (`r unique(myanswer3$fungus)`). `r sum(myanswer3$foldchange_mean > 0)` genes were upregulated and `r sum(myanswer3$foldchange_mean < 0)` were downregulated.


## Problem 5

Plot the expression (`value`) by time (`hpi`) for the above genes. Format the plot as follows:

1. Plot each point.
1. Connect the points with a smooth line 
1. Create a facet for each `plant` / `bgh` pair 
1. Assign a different color to each gene 
1. Assign a different shape to each replicate 
1. Add a descriptive title to the plot 
1. Use the "black & white" theme

Extra credit: add error bars to the plot (use `geom_errorbar`).

```{r problem_5}
# is there a dplyr way to filter one tibble by another tibble?
myanswer5 <- merge(tidy_foldchange, myanswer3, by = c('gene.name', 'plant', 'fungus'))[1:7]
myanswer5$time.hpi.x <- as.numeric(gsub("hpi\\.", "", myanswer5$time.hpi.x))



myplot <- ggplot(myanswer5, aes(x = time.hpi.x, y = foldchange_mean.x, colour = gene.name))
myplot <- myplot + geom_point() + geom_smooth(method = 'lm', se = FALSE) + facet_grid(plant ~ fungus) + theme_bw()
myplot <- myplot + labs(x = "Hours Post Infection", y = "Transcript Foldchange", title = 'B. graminis gene expression during Arabidopsis infections')
myplot <- myplot + geom_errorbar(aes(ymin = foldchange_mean.x-foldchange_var.x/2, ymax = foldchange_mean.x+foldchange_var.x/2))
myplot <- myplot + ylim(-20,20)


print(myplot)
```
